Vending step5

AppleCEO opened the pull request:
요구사항 반영했습니다.

- viewDidLoad 에서 Observe 등록
- 모델 객체에 상태가 변화하는 메소드마다 NotificationCenter에 post

AppleCEO committed:
수정 - 메소드명 조금 더 정확하게 작성

AppleCEO committed:
추가 - 옵저버 등록, 실행할 함수 선언

AppleCEO committed:
추가 - 모델에 상태 변화하는 메소드에 옵저버 post 추가

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -8,6 +8,11 @@
+extension Notification.Name {
+    static let addStock = Notification.Name("addStock")
+    static let addBalance = Notification.Name("addBalance")
+}
```
이 노티피케이션은 누가 만들어내고 보내는 걸까요?
그렇다면 노티피케이션 정의는 어디에 있는게 좋을까요?

AppleCEO replied:
노티피게이션은 모델에서 보내기 때문에 VendingMachine 파일로 이동하도록 하겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachine.swift:
```diff
@@ -135,6 +139,9 @@ final class VendingMachine: VendingMachineManagementable, VendingMachineUseable,
         try balance.minusBalance(drink.getPrice())
+        
+        NotificationCenter.default.post(name: .addStock, object: nil)
+        NotificationCenter.default.post(name: .addBalance, object: nil)
```
음료를 구매하는 동작에서 .addStock 이나 .addBalance 라는 의미가 맥락에 안 맞는 것 같습니다.

AppleCEO replied:
.refreshStock, .refreshBalance 라고 바꿔봤습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -54,12 +61,20 @@ class ViewController: UIViewController {
     }
     
     @IBAction func insertCoin(_ sender: UIButton) {
         vendingMachine.insertCoin(sender.tag)
```
sender.tag 를 매개변수로 직접 넘기는 것은 매우 위험한 코드입니다. 
다른 사람이 View에서 tag 값을 잘 못 넣거나 Action을 새로 연결하고 값을 바꾸지 않아도 오류가 생길 수 있습니다.
직접 값을 활용하기 보다는 검증이나 매칭, 데이터 타입으로 연결해서 사용하는 방식이 좋습니다.

AppleCEO replied:
if 로 검증하도록 구성했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -54,12 +61,20 @@ class ViewController: UIViewController {
 
```
refreshDrinkCount() 나 refreshBalance() 에서도 음료수 전체나 잔액을 직접 가져오기 보다는
뷰에 출력을 위한 프로토콜을 만들어서 값을 그대로 가져오지 않지만 채울 수 있도록 만드는 게 좋습니다. 
이전 스탭에서도 OutputView 출력할 때 그대로 모두 출력했었나요?

AppleCEO replied:
예전 스텝에도 그대로 모두 출력했었습니다. 조금 더 힌트 주실 수 있나요?

godrm replied:
Delegate 나 DataSource 프로토콜처럼 출력을 위한 프로토콜을 선언해서
전체 값을 가져오지 않고, 프로토콜의 메소드를 호출해서 출력할 수 있는 방법을 생각해보세요.

AppleCEO replied:
아래 내용으로 수정해봤습니다.

Model 에서 데이터가 바뀌면 Notification Center 을 통해 controller 로 출력을 위한 데이터(전체 값이 아닌 데이터)를 전송합니다.
Controller 에서 받은 데이터를 View 에 반영합니다.

godrm replied:
모델 객체에서 notify를 보내는 구조는 좋은 구조입니다.
다만 그 시점을 controller가 직접 호출하는 거라면 좋은 방식 구현이 아닙니다. 
오히려 값을 보내려다보니 구체적인 타입이었던 Money 정보도 없애버렸군요. 
...
뷰 -> 뷰 컨트롤로 액션에서는 모델에 값을 변경하기 위한 동작만 처리한다.
모델에서는 동작에 따라 값이 변경되면 (누군가에게) 변경된 값을 통보한다. 
뷰 컨트롤러는 통보를 확인하고 뷰를 업데이트한다. 까지만 정리합시다. 바로 이전 피드백은 무시하세요.

godrm replied:
제가 하고 싶었던 말은 ViewController 가 VendingMachine 에서 값을 가져오지 말고,
출력에 필요한 프로토콜을 만들어서 VendingMachine에서 채택하고 프로토콜을 기반으로 뷰에 값을 지정하도록 구현해보자는 거였습니다. 이런 방식으로 구현 안했었다면 꼭 해봐야합니다.

AppleCEO replied:
출력에 필요한 프로토콜(StockPrintable, BalancePrintable) 을 생성했습니다.
VendingMachine 에서 출력을 위한 프로토콜을 채택하도록 구성했습니다.
기존에 컨트롤러에서 모델의 notify 메서드를 직접 호출하던 소스는 제거하였습니다.

godrm reviewed (commented):
다음 피드백을 확인해보세요.

AppleCEO committed:
이동 - Notification.Name 의 위치를 호출하는 모델 파일로 이동

AppleCEO committed:
수정 - refresh 로 시작하도록 옵저버 이름 변경

AppleCEO committed:
수정 - tag 값이 정해진 범위에 맞는지 확인 추가

AppleCEO committed:
수정 - notify 할 때 데이터도 보내도록 수정

기존에 모델에서 notificationCenter 로 데이터가 바뀌었다는 걸 Controller 에게 알려주면 Contorller 에서 Model 의 데이터를 가져오는 메소드를 실행하던 것 -> 모델이 변경되면 변경된 값을 Controller 에게 보내주도록 수정

AppleCEO committed:
수정 - 의미없이 한 칸 띄어있던 것 제거

AppleCEO committed:
수정 - 전체 값을 가져오지 않고 출력을 위한 데이터를 만들어 전송

AppleCEO committed:
생성 - 하드코딩된 한글을 enum 의 rawValue 로 구현

AppleCEO committed:
생성 - 자판기 재고를 출력할 수 있는 프로토콜 생성

AppleCEO committed:
생성 - 자판기 잔고를 출력할 수 있는 프로토콜 생성

AppleCEO committed:
추가 - 출력을 위한 프로토콜 채택

AppleCEO committed:
수정 - 컨트롤러에서 모델의 notify 를 실행하던 소스 제거

AppleCEO committed:
변경 - 프로토콜 생성에 대한 config 파일 변경
